[MRG + 1] Fix sample_without_replacement speed issue

jkarno opened the pull request:
#### Reference Issue

Fixes #7091 
#### What does this implement/fix? Explain your changes.

This replaces all calls to sample_without_replacement with an np.permutation call, as per discussion in issue thread.
#### Any other comments?
1. Running the tests, this doesn't currently seem to pass but I'm not exactly sure what I'm misunderstanding. Does it have to do with seeding on the random states or otherwise incorporating the random variables?
2. Should I remove the files implementing sample_without_replacement, or is does this fix just involve replacing the existing calls with np.permutation?

jkarno committed:
WIP replace sample_without_replacement

jkarno committed:
WIP fix deterministic state sample_without_replacement

jkarno commented:
@amueller Made the changes to deterministic random variables but the Travis PR still seems to be error-ing. Are this specific steps I need to take when I submit code changes that I might be missing?

nelson-liu commented:
@jkarno I downloaded your PR and ran the tests on my machine, and it hung for a long time / took up a lot of processing power (according to htop) on one of the PCA tests. Can you reproduce the behavior I saw when running tests on your local machine? It seems like travis automatically killed the job for some reason; high usage would be a reasonable one...

nelson-liu commented:
ah yes it was [sklearn.decomposition.tests.test_pca.test_explained_variance](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/decomposition/tests/test_pca.py#L194)

jkarno commented:
Yeah, I do notice it hanging on that one too. I'll take a look and try to figure out what could be causing that, it seems strange that these changes are affecting that test since I don't immediately see how they are connected.

jkarno commented:
Found the connection: That test calls make_classification, which calls generate_hypercube, which uses permutation.
`out = astype(random_state.permutation(2 ** dimensions)[:samples],
                 dtype='>u4', copy=False)`

Trying to figure out now why this case is so slow, but it might be a case with a very large number being passed in, due to the exponential.

jkarno committed:
remove unecessary step for sample_without_replacement fix

jkarno commented:
Last update for now - I realized that check_random_state was unnecessary in the generate_hypercube method since it will be called from a method that does that already. Removing it fixed the hanging on my machine, but that same test now errors instead of hangs.

Committing that change to see if you guys have any thoughts.

nelson-liu commented on sklearn/datasets/samples_generator.py:
```diff
@@ -28,8 +27,7 @@ def _generate_hypercube(samples, dimensions, rng):
                           _generate_hypercube(samples, 30, rng)])
-    out = astype(sample_without_replacement(2 ** dimensions, samples,
-                                            random_state=rng),
+    out = astype(random_state.permutation(2 ** dimensions)[:samples],
```
seems that the issue is that random_state is now undefined here?

amueller replied:
should be rng

amueller replied:
I highly recommend using an editor that supports pyflakes or flake8 to catch this sort of error while writing the code.

jkarno replied:
Sorry for the delayed response, but even with that fix I can't seem to figure out the issue with the test case hanging. Is it possible that the permutation method is actually slower in this case for some reason?

amueller replied:
That seems unlikely. Have you tried setting a breakpoint?

amueller replied:
Which test by the way?

jkarno replied:
This test: https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/decomposition/tests/test_pca.py#L194

Which I found the connection to this method: https://github.com/scikit-learn/scikit-learn/pull/7703#issuecomment-255006786

amueller replied:
Yeah this calls it with ``N=302231454903657293676544``. Sooo... we either limit hypercube to a smaller dimension or we make ``auto`` default to our implementation in the very strong edge-cases.

jkarno replied:
So should it go back to calling "sample_without_replacement" and then have auto determine whether to use permutation or not, or should it just make that determination directly in this one case and only call sample_without_replacement in the edge cases?

amueller replied:
go back to calling ``sample_without_replacement`` and then use the heuristic I gave to decide whether to use permutation. thanks :)

amueller commented:
@jkarno any chance you can work on this today or tomorrow? I would really like to have this in the 0.18.1 release.

jkarno committed:
add sample_without_replacement permutation condition

amueller commented:
Can you also please post the output of ``python benchmarks/bench_sample_without_replacement.py --n-step 20 --n-times 50``?

jkarno commented:
![benchmark_results](https://cloud.githubusercontent.com/assets/5368880/19703184/e64bdae0-9ad0-11e6-9d5c-07d1c5b2b731.png)

jkarno commented:
![benchmark_results](https://cloud.githubusercontent.com/assets/5368880/19703386/9f678db2-9ad1-11e6-8aa1-a9fe9b0bbcb8.png)

jnothman commented:
It looks like perhaps we should be using tracking selection for more of the head and tail.

Also would be worth benchmarking other versions of numpy.

jkarno commented:
I haven't tried other versions of numpy yet, but here are the results with only using permutation between 0.15 and 0.75.
![benchmark_results](https://cloud.githubusercontent.com/assets/5368880/19704299/2ce3f844-9ad5-11e6-8eb9-13c33c4e4e03.png)

amueller commented:
travis is failing. What numpy are you using, and are you using mkl?

jkarno commented:
I currently have numpy version 1.9.2 (which I see now is outdated), and I don't believe I'm using mkl

amueller commented:
on my machine the break-even point is much closer to 0.01 and 0.99 (see what I posted in the other thread).

amueller commented:
In the other thread, @arjoly pointed out there was a big difference with numpy 0.11

jkarno commented:
Okay I'll update numpy and try again. Also, the tests that are failing on this branch and not on my master branch seem to be "doctests". I'm not exactly sure why this change is causing them to fail but I'll keep looking into it.

jkarno commented:
Seeing much better results with numpy 0.11.2 now and the 0.01 and 0.99 thresholds.
![benchmark_results](https://cloud.githubusercontent.com/assets/5368880/19705969/75edaabe-9add-11e6-81a1-d9ef36cd42dc.png)

jnothman commented:
Looking at the code, I see that tracking is overridden with reservoir if ratio >= 0.2. It's a bit weird that this is not documented.

jnothman commented:
So how do we decide this issue?

My main concern is that we do this once and it not be variable under changes of Numpy version, to the extent that this is possible, so as not to affect estimator models more often than necessary (affected estimators: RANSAC, Bagging*, RandomizedSearchCV, sparse RandomProjection, DummyClassifier with sparse output; also `make_classification`). Did the reimplementation of `numpy.permutation` in np 0.11 produce the same output as the previous version anyway?

The best option may be just to throw old numpy users to the wind and adopt numpy.permutation in (0.01, 0.99).

amueller commented:
I think the "to the wind" is actually not that bad. It's still faster than the current implementation and it's pretty close to optimal everywhere. (only in (0.01,1) and (0.90, 99) there is a difference).
So I'd keep it like it is implemented right now, and document the function properly. We can check whether RandomState(0).permutation gives the same result before or after the rewrite but I feel that is outside of our responsibility.

@jkarno the doctests fail because the randomness is somewhat different now. Just change the expected outcome in these places.

jkarno committed:
update doctests and documentation for sample_without_replacement

jkarno commented:
Fixed the doctests and added some documentation. Looks like it's passing Travis now.

@amueller One issue I'm having on my personal machine currently is it seems to be compiling with the wrong version of numpy and then reporting an error that "numpy.dtype has the wrong size" and exiting. I tried uninstalling and reinstalling numpy, pandas, scipy, and then scikit-learn but I can't seem to fix it. Here is the traceback:

> File "/Library/Python/2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/Library/Python/2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/Library/Python/2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/scikit-learn/sklearn/__init__.py", line 57, in <module>
    from .base import clone
  File "/scikit-learn/sklearn/base.py", line 12, in <module>
    from .utils.fixes import signature
  File "/scikit-learn/sklearn/utils/__init__.py", line 10, in <module>
    from .murmurhash import murmurhash3_32
  File "__init__.pxd", line 155, in init sklearn.utils.murmurhash (sklearn/utils/murmurhash.c:6304)
ValueError: numpy.dtype has the wrong size, try recompiling. Expected 88, got 96

amueller commented:
@jkarno how did you install numpy, scipy and scikit-learn?
There are probably multiple python environments on your machine. scikit-learn is currently using the system python.
I would recommend that you either create a vritual environment and use that for your python programming or install conda and exclusively use that environment. I think using the system python on OS X leads to trouble (though I have no personal experience with it).

amueller commented:
LGTM

jkarno committed:
small doc change sample_without_replacement

jnothman commented:
Could you please write a what's new entry, noting that samples may differ from previously for a fixed random state, and which estimators this affects...

GaelVaroquaux commented:
> I think the "to the wind" is actually not that bad. It's still faster than the
> current implementation and it's pretty close to optimal everywhere. (only in
> (0.01,1) and (0.90, 99) there is a difference).

So let's do this!

jnothman commented:
It looks like perhaps you're not using the installed scikit-learn when
running tests, but a local working copy. Try running `make in`

On 27 October 2016 at 06:53, Josh Karnofsky <notifications@github.com>
wrote:

> Fixed the doctests and added some documentation. Looks like it's passing
> Travis now.
>
> @amueller <https://github.com/amueller> One issue I'm having on my
> personal machine currently is it seems to be compiling with the wrong
> version of numpy and then reporting an error that "numpy.dtype has the
> wrong size" and exiting. I tried uninstalling and reinstalling numpy,
> pandas, scipy, and then scikit-learn but I can't seem to fix it. Here is
> the traceback:
>
> File "/Library/Python/2.7/site-packages/nose/loader.py", line 418, in
> loadTestsFromName
> addr.filename, addr.module)
> File "/Library/Python/2.7/site-packages/nose/importer.py", line 47, in
> importFromPath
> return self.importFromDir(dir_path, fqname)
> File "/Library/Python/2.7/site-packages/nose/importer.py", line 94, in
> importFromDir
> mod = load_module(part_fqname, fh, filename, desc)
> File "/scikit-learn/sklearn/*init*.py", line 57, in
> from .base import clone
> File "/scikit-learn/sklearn/base.py", line 12, in
> from .utils.fixes import signature
> File "/scikit-learn/sklearn/utils/*init*.py", line 10, in
> from .murmurhash import murmurhash3_32
> File "*init*.pxd", line 155, in init sklearn.utils.murmurhash
> (sklearn/utils/murmurhash.c:6304)
> ValueError: numpy.dtype has the wrong size, try recompiling. Expected 88,
> got 96
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7703#issuecomment-256458452>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69F5Na7q50WzI1CjO2GMmZ-iU1pDks5q36_QgaJpZM4Kak8K>
> .
>

jkarno commented:
@jnothman I'm writing up the what's new entry now, but I'm not sure exactly how to determine all of the estimators that are affected. I went through the files that reference sample_without_replacement, but I'm not sure if some of the methods are indirectly used (like through the samples_generator methods). I currently have determined that RandomizedSearchCV, RANSACRegressor, BaggingClassifier, BaggingRegressor are affected.

amueller commented:
the sample_generators are only used for examples. You should document that it changed, but that doesn't have an impact on anything else.
I think you got everything but ``SparseRandomProjection``. Do a ``git grep sample_without_replacement``

jkarno committed:
add sample_without_replacement documentation to whats_new

jnothman commented:
Thanks, @jkarno!
